packages/components/src/navigator/router/memory-router.js

```diff
@@ -0,0 +1,33 @@
+/**
+ * External dependencies
+ */
+import { createMemoryHistory as createHistory } from 'history';
+
+/**
+ * Internal dependencies
+ */
+import Router from './router';
+
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+
+/**
+ * The public API for a <Router> that stores location in memory.
+ */
+class MemoryRouter extends Component {
+	constructor() {
+		super();
+
+		this.history = createHistory( this.props );
+	}
+
+	render() {
+		return (
+			<Router children={ this.props.children } history={ this.history } />
+		);
+	}
+}
```

Comment:
We could avoid a class component by using a closure and doing something similar to `react-nagivation`:

```jsx
const createMemoryRouter = () => {
  let history;

  return function MemoryRouter(props) {
    history = createHistory(props);
    return <Router children={this.props.children} history={history} />;
  };
}
```